README.md
EcoNetGen
================

[![Travis-CI Build
Status](https://travis-ci.org/cboettig/EcoNetGen.svg?branch=master)](https://travis-ci.org/cboettig/EcoNetGen)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/cboettig/EcoNetGen?branch=master&svg=true)](https://ci.appveyor.com/project/cboettig/EcoNetGen)
[![Coverage
Status](https://img.shields.io/codecov/c/github/cboettig/EcoNetGen/master.svg)](https://codecov.io/github/cboettig/EcoNetGen?branch=master)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/EcoNetGen)](https://cran.r-project.org/package=EcoNetGen)
[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![downloads](https://cranlogs.r-pkg.org/badges/grand-total/netgen)](https://github.com/metacran/cranlogs.app)
[![DOI](https://zenodo.org/badge/116610054.svg)](https://zenodo.org/badge/latestdoi/116610054)

<!-- README.md is generated from README.Rmd. Please edit that file -->

`EcoNetGen` lets you randomly generate a wide range of interaction
networks with specified size, average degree, modularity, and
topological structure. You can also sample nodes and links from within
simulated networks randomly, by degree, by module, or by abundance.
Simulations and sampling routines are implemented in ‘FORTRAN’,
providing efficient generation times even for large networks. Basic
visualization methods also included. Algorithms implemented here are
described in de Aguiar et al. (2017)
[arXiv:1708.01242](https://arxiv.org/abs/1708.01242).

## Installation

`EcoNetGen` is now on CRAN and can be installed in the usual way:

``` r
install.packages("EcoNetGen")
```

See [NEWS](NEWS.md) for a list of the most recent changes to the
development version and current CRAN release. You can install the
current development version of `EcoNetGen` from GitHub with:

``` r
# install.packages("devtools")
devtools::install_github("cboettig/EcoNetGen")
```

This way requires you have a recent FORTRAN compiler avialble on your
machine.

## Randomly generate networks

This is a basic example which generates a network. See `?netgen` for
documentation describing the parameter arguments. Setting `verbose =
FALSE` (default) surpresses the output summary message.

``` r
library(EcoNetGen)
set.seed(123456) # for a reproducible simulation

network <- netgen(net_size = 150,
                  ave_module_size = 20, 
                  min_module_size = 10,
                  min_submod_size = 5,
                  net_type = "bi-partite nested",
                  ave_degree = 10,
                  verbose = TRUE
                  ) 
#> 
#> module count = 8 
#> average degree = 6.08 
#> average module size = 18.75 
#> number of components = 1 
#> size of largest component = 150
```

We can plot the resulting `igraph` as an adjacency matrix:

``` r
adj_plot(network)
```

![](man/figures/README-adj_plot-1.png)<!-- -->

Network `igraph` objects can also be plotted using the standard `igraph`
plotting routines, for example:

``` r
library(igraph)
plot(network, vertex.size= 0, vertex.label=NA, 
     edge.color = rgb(.22,0,1,.02), vertex.shape="none", 
     edge.curved =TRUE, layout = layout_with_kk)
```

![](man/figures/README-igraph_plot-1.png)<!-- -->

## Sample from a network

``` r
set.seed(123456) # for a reproducible random sampling
sampled <- netsampler(network, 
                     key_nodes_sampler = "degree", 
                     neighbors_sampler = "random",
                     n_key_nodes = 50,
                     n_neighbors = 0.5 # 50%
                    )
```

We can plot the adjacency network, coloring red the sampled nodes. Note
that `adj_plot` objects are just `ggplot` graphs (`geom_raster`) under
the hood, and can be modified with the usual `ggplot` arguments, such as
adding a title and changing the color theme here.

``` r
library(ggplot2) # needed to modify plot

adj_plot(sampled) + 
  ggtitle("Adjacency matrix of sampled vs full network") + 
  scale_fill_manual(values = c("#ED4E33", "#3B7EA1"))
```

![](man/figures/README-sampled_adj-1.png)<!-- -->

Don’t forget to check out the `ggraph` package, which isn’t required for
`EcoNetGen` but provides a lot of additional great ways to plot your
network. Here we plot the simulated network color-coding the sampled
nodes and edges (indicated by the label “sampled” on vertices and
edges):

``` r
library(ggraph)
ggraph(sampled, layout = 'kk') +
        geom_edge_link(aes(color = label), alpha=0.4) +
        geom_node_point(aes(color = label)) +
        theme_graph() + 
        scale_color_manual(values = c("#ED4E33", "#3B7EA1")) + 
        scale_edge_color_manual(values = c("#ED4E33", "#3B7EA1"))
```

![](man/figures/README-sampled_hairball-1.png)<!-- -->

Or extract and plot just the sampled network:

``` r
subnet <- subgraph.edges(sampled, 
                         E(sampled)[label=="sampled"])

ggraph(subnet, layout = 'graphopt') +
        geom_edge_link(alpha=0.4) +
        geom_node_point() +
        theme_graph() 
```

![](man/figures/README-subgraph-1.png)<!-- -->

## A few example statistics

And we can compute common statistics from igraph as well. Here we
confirm that clustering by “edge betweeness” gives us the expected
number of modules:

``` r
community <- cluster_edge_betweenness(as.undirected(network))
length(groups(community))
#> [1] 9
```

We can check the size of each module as well:

``` r
module_sizes <- sizes(community)
module_sizes
#> Community sizes
#>  1  2  3  4  5  6  7  8  9 
#> 18 19 31 19 22  7  3 20 11
```

Average degree:

``` r
mean(degree(as.undirected(network)))
#> [1] 6.08
```

We can also label and plot the cluster membership:

``` r
V(sampled)$module <- as.character(membership(community))
```

``` r
ggraph(sampled, layout = 'kk') +
  geom_edge_link(alpha=0.1) +
  geom_node_point(aes(colour = module)) + 
  theme_graph()
```

![](man/figures/README-members-1.png)<!-- -->


NEWS.md
# NetGen 0.2.2

* Avoid potential crashes from FORTRAN code by intercepting any
  invalid parameter combinations and returning a helpful error message.
  Please note, this changes the default parameter settings of `min_module_size`
  and `ave_degree`, which involved a potentially combination for scale-free
  and thus also some mixed-type networks.  
  
  
  
  

# NetGen 0.2.1

* fix inappropriate default in `netgen_v1` routine, and add assertion to check argument lengths.

# NetGen 0.2.0

* Add `netsampler()` routine

* Modify user interface for `netgen()` to provide more self-explanatory argument names. Note: the 
  old behaviour can be recovered for previous scripts you don't want to update manually by doing 
  `netgen <- EcoNetGen:::netgen_v1` after loading `EcoNetGen`.
  
* Ensure that FORTRAN correctly obeys `set.seed()` by always using the R random number generator. [#6](https://github.com/cboettig/EcoNetGen/issues/6)

# NetGen 0.1.1 2018-04-10

* Initial release to CRAN

# NetGen 0.1.0 2018-04-04

* Added a `NEWS.md` file to track changes to the package.





tests/testthat.R
library(testthat)
library(EcoNetGen)

## Windows i386 builds always time out when ggraph
## circle layout is present in README images.  CRAZY
skip <- (R.Version()$arch == "i386") &
  "windows" %in% tolower(Sys.info()[["sysname"]])

if(!skip){
  test_check("EcoNetGen")
}


tests/testthat/test-netgen.R
library(testthat)

context("netgen")
set.seed(12345)

test_that("we can create a random type network", {
  #skip_if(R.Version()$arch == "i386")

  net <- netgen()
  expect_is(net, "igraph")

})


test_that("testing error messages", {

  expect_error(netgen(min_module_size = 1))

  expect_error(netgen(net_type = "mixed",
                      ave_degree = 10,
                      ave_module_size = 9))

  expect_error(netgen(min_module_size = 0))
  expect_error(netgen(min_submod_size = 0))
  expect_error(netgen(rewire_prob_global = 0))
  expect_error(netgen(rewire_prob_global = 10))
  expect_error(netgen(rewire_prob_local = 10))
  expect_error(netgen(mixing_probs = 10))

})


netgen_v1 <- EcoNetGen:::netgen_v1

test_that("we can create a random type network (v1)", {
  #skip_if(R.Version()$arch == "i386")

  net <- netgen_v1()
  expect_is(net, "igraph")

})

test_that("we can create a mixed type network", {

  n_modav <- c(150,30)
  cutoffs <- c(15,5)
  net_type <- 0
  net_degree <- 10.0
  net_rewire <- c(0.07,0.2)
  mod_probs <- c(0.2, 0.0, 0.3, 0.3, 0.2, 0.0, 0.0)

  M <- netgen_v1(n_modav,cutoffs,net_type,net_degree,net_rewire,mod_probs)

  expect_is(M, "igraph")
  p1 <- adj_plot(M)
  expect_is(p1, "ggplot")

  })


test_that("we can create all types of network", {

        n_modav <- c(150,20)
        cutoffs <- c(15,5)
        net_type <- 0
        net_degree <- 10.0
        net_rewire <- c(0.07,0.2)
        mod_probs <- c(0.2, 0.0, 0.3, 0.3, 0.2, 0.0, 0.0)

        M <- netgen_v1(n_modav,cutoffs,1,net_degree,net_rewire,mod_probs)
        expect_is(M, "igraph")
        M <- netgen_v1(n_modav,cutoffs,2,net_degree,net_rewire,mod_probs)
        expect_is(M, "igraph")
        M <- netgen_v1(n_modav,cutoffs,3,net_degree,net_rewire,mod_probs)
        expect_is(M, "igraph")
        M <- netgen_v1(n_modav,cutoffs,41,net_degree,net_rewire,mod_probs)
        expect_is(M, "igraph")
        M <- netgen_v1(n_modav,cutoffs,51,net_degree,net_rewire,mod_probs)
        expect_is(M, "igraph")
        M <- netgen_v1(n_modav,cutoffs,42,net_degree,net_rewire,mod_probs)
        expect_is(M, "igraph")
        M <- netgen_v1(n_modav,cutoffs,52,net_degree,net_rewire,mod_probs)
        expect_is(M, "igraph")

})

test_that("setting seed creates a reproducible network", {

  set.seed(5555555)
  M <- netgen_v1()
  set.seed(5555555)
  M2 <- netgen_v1()

  expect_identical(igraph::as_adj(M), igraph::as_adj(M2))

})





tests/testthat/test-netsample.R
library(testthat)
testthat::context("netsampler")

set.seed(12345)
netgen_v1 <- EcoNetGen:::netgen_v1

testthat::test_that("we can run netsampler", {

  library(EcoNetGen)
  net <- netgen_v1()
  testthat::expect_is(net, "igraph")
  res <- netsampler(net)
  testthat::expect_is(res, "igraph")

  p1 <- adj_plot(res)
  testthat::expect_is(p1, "ggplot")
})


testthat::test_that("net sample works with other settings", {

  network_in <- netgen_v1()
  sn <- netsampler(network_in,
                 key_nodes_sampler = "lognormal",
                 neighbors_sampler = "exponential")
  testthat::expect_is(sn, "igraph")

  sn <- netsampler(network_in,
                  n_neighbors = 3)
  testthat::expect_is(sn, "igraph")

  sn <- netsampler(network_in,
                  key_nodes_sampler = "Fisher")
  testthat::expect_is(sn, "igraph")

  sn <- netsampler(network_in,
                  key_nodes_sampler = "exponential")
  testthat::expect_is(sn, "igraph")

  sn <- netsampler(network_in,
                  key_nodes_sampler = "degree")
  testthat::expect_is(sn, "igraph")

  sn <- netsampler(network_in, "module")
  testthat::expect_is(sn, "igraph")

})

testthat::test_that("we can run netsample",{

      library(EcoNetGen)
      library(igraph)

      network_in <- netgen_v1()
      out <- netsampler(network_in)

      ## Subset the sampled network:
      sampled <- igraph::subgraph.edges(out, E(out)[label=="sampled"])

      testthat::expect_is(out, "igraph")
      testthat::expect_is(sampled, "igraph")



})





src/zzz.c
#include <R.h>
#include <R_ext/Rdynload.h>
#include <Rversion.h>

void F77_SUB(rndstart)(void) { GetRNGstate(); }
void F77_SUB(rndend)(void) { PutRNGstate(); }
double F77_SUB(unifrnd)(void) { return unif_rand(); }

void F77_SUB(subnetgen)(int *, int *, int *, int *,float *, float *, float *, int *);
void F77_SUB(clusters)(int *, int *, int *, int *);
void F77_SUB(subsampling)(int *, int *, int *, int *, float *,
                          int *, int *, int *, int *, int *, int *, int *);

static R_FortranMethodDef R_FortranDef[] = {
  {"rndstart",  (DL_FUNC) &F77_SUB(rndstart),  0, NULL},
  {"rndend",  (DL_FUNC) &F77_SUB(rndend),  0, NULL},
  {"unifrnd",  (DL_FUNC) &F77_SUB(unifrnd),  0, NULL},
  {"subnetgen",  (DL_FUNC) &F77_SUB(subnetgen),  8, NULL},
  {"clusters",  (DL_FUNC) &F77_SUB(clusters),  4, NULL},
  {"subsampling",  (DL_FUNC) &F77_SUB(subsampling),  12, NULL},

  {NULL, NULL, 0, NULL}
};


void R_init_NetGen(DllInfo *dll) {
  R_registerRoutines(dll, NULL, NULL, R_FortranDef, NULL);
#if defined(R_VERSION) && R_VERSION >= R_Version(3, 3, 0)
  R_useDynamicSymbols(dll, FALSE);
#endif
}


R/NetPlot.R
# See https://matthewlincoln.net/2014/12/20/adjacency-matrix-plots-with-r-and-ggplot2.html
# Also, see matrix plot method in dev version of ggraph


#' Plot network adjacency matrix
#'
#' @param graph an igraph object
#' @export
#' @importFrom ggplot2 aes_string ggplot scale_x_discrete scale_y_discrete
#' @importFrom ggplot2 element_blank theme_bw xlab ylab theme geom_raster
#' @importFrom igraph get.data.frame
#' @examples
#' \donttest{
#' set.seed(12345)
#' graph <- netgen()
#' adj_plot(graph)
#' }
#'
adj_plot <- function(graph){

  edge_list <- igraph::get.data.frame(graph, what = "edges")

  if("label" %in% names(edge_list)){
    g <- ggplot(edge_list, aes_string(x = "from", y = "to", fill="label"))
  } else {
    g <- ggplot(edge_list, aes_string(x = "from", y = "to"))
  }
      g +
      geom_raster() + theme_bw() +
      scale_x_discrete(drop = FALSE) +
      scale_y_discrete(drop = FALSE) +
      xlab("") +
      ylab("") +
      theme(
        axis.text = element_blank(),
        aspect.ratio = 1,
        legend.position = "none",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks = element_blank())
}




inst/crash.R
library(EcoNetGen)
set.seed(2222)

## Runs just fine
replicate(1000,
netgen(net_size = 50,
       ave_module_size = 10,
       min_module_size = 4,
       min_submod_size = 2,
       net_type = "bi-partite nested",
       ave_degree = 10,
       rewire_prob_global = 0.3,
       rewire_prob_local = 0.1,
       mixing_probs = c(0.2, 0.2, 0.2, 0.2, 0.2, 0.2 ,0.2),
       verbose = FALSE)
)


set.seed(1234)
## Catches error:
          netgen(net_size = 30,
                 ave_module_size = 10,
                 min_module_size = 1,
                 min_submod_size = 1,
                 net_type = "bn",
                 ave_degree = 10,
                 rewire_prob_global = 0.2,
                 rewire_prob_local = 0.0,
                 mixing_probs = c(0.2, 0.2, 0.2, 0.2, 0.2, 0.0 ,0.0),
                 verbose = FALSE)




set.seed(2222)
replicate(1000,
          netgen(net_size = 50,
                 ave_module_size = 25,
                 min_module_size = 10,
                 min_submod_size = 1,
                 net_type = "mixed",
                 ave_degree = 9,
                 rewire_prob_global = 0.0,
                 rewire_prob_local = 0.0,
                 mixing_probs = c(1,1,1,0.1,0.1,0.1,0.1),
                 verbose = FALSE)
)





R/NetSampler.R
#' Network Sampling Routine
#'
#' @param network_in input network (as igraph object)
#' @param key_nodes_sampler sampling criteria for key nodes. See details.
#' @param neighbors_sampler sampling criteria for neighbors. see details.
#' @param n_key_nodes number of key nodes to sample.
#' @param n_neighbors number of first neighbors or fraction of first neighbors.
#'  See details.
#' @param hidden_modules list of the modules to exclude
#' (max 10 modules; only the first numb_hidden are used)
#' @param module_sizes integer vector giving the size of each module. see details.
#' @param cluster_fn a clustering function, from `igraph::cluster_*`. Default is
#' `igraph::cluster_edge_betweeness`.  Only used to compute module sizes if not
#' provided.
#' @details
#' Algorithm first samples n_key_nodes according the the requested `key_nodes_sampler`
#' criterion.  For each key node, the requested number or fraction of neighbors is
#' then sampled according to the `neighbors_sampler` criterion.  Optionally, a list of
#' modules can be designated as "hidden" and will be excluded from sampling.
#'
#' if `n_neighbors is greater than 1, assumes this is the number to sample.  If
#' `n_neighbors` is between 0 and 1, assumes this is the fration of neighbors to
#' sample.  (To sample 1 neighbor, use an explicit integer, `1L` (or as.`integer(1)`)
#' to sample 100% of neighbors, use a numeric, 1.0.
#'
#' Provide `module_sizes` list to improve performance.  If not provided, this will
#' will be calculated based on `igraph::cluster_edge_betweeness`.  Be sure to
#' provide a `module_sizes` vector whenever calling `netsampler` repeatedly on the
#' same network to avoid unnecessary performance hit from recalculating modules every
#' time.  See examples.
#'
#' @return the original input network (as an igraph network object),
#'  with the attribute `label` added to the edges and vertices indicating
#' if that edge or vertex was `sampled` or `unsampled`.
#'
#' @export
#' @importFrom igraph cluster_edge_betweenness as.undirected groups as_adjacency_matrix
#' delete_edge_attr E sizes
#'
#' @examples
#' \donttest{
#' set.seed(12345)
#' net <- netgen()
#' sample <- netsampler(net)
#'
#' ## Precompute `module_sizes` for replicate sampling of the same network:
#'  library(igraph)
#'  modules <- cluster_edge_betweenness(as.undirected(net))
#'  module_sizes <- vapply(igraph::groups(modules), length, integer(1))
#'  sample <- netsampler(net, module_sizes = module_sizes)
#'
#' }
#'
netsampler <-
  function(network_in,
           key_nodes_sampler = c("random", "lognormal", "Fisher log series",
                               "exponential", "degree", "module"),
           neighbors_sampler = c("random", "exponential"),
           n_key_nodes = 10,
           n_neighbors = 0.5,
           hidden_modules = NULL,
           module_sizes = NULL,
           cluster_fn = igraph::cluster_edge_betweenness
           ) {

    key_nodes_sampler <- match.arg(key_nodes_sampler)
    neighbors_sampler <- match.arg(neighbors_sampler)

    crit <- c(which(c("random", "lognormal", "Fisher log series",
                      "exponential", "degree", "module") ==
                    key_nodes_sampler) - 1,
              which(c("random", "exponential") == neighbors_sampler) - 1
    )

    #initial, final, delta,  final & delta not used.
    key_nodes <- c(n_key_nodes, 0, 0, 0)

    if(is.integer(n_neighbors)){
      if(n_neighbors == 1L)
        n_neighbors <- 1.1
    }

    anfn <- n_neighbors

    numb_hidden <- length(hidden_modules)

    ## NOTE: if module_sizes are not provided, then we calculate
    ## module sizes on the fly, though the FORTRAN netgen code
    ## returns these. Not clear what clustering algorithm
    ## Marcus's FORTRAN routine uses for that. This has some
    ## additional performance overhead; in particular, code
    ## seeking to create many resamplings from the same network
    ## should be sure to include module sizes list.

    if(is.null(module_sizes)){
      community <- cluster_fn(
        igraph::as.undirected(network_in))
      module_sizes <- igraph::sizes(community)
    }

    ## Convert igraph to integer vector
    net <-
      as.integer(
        as.matrix(igraph::as_adjacency_matrix(network_in)))

    ## number of nodes
    n <- as.integer(sqrt(length(net)))

    res <- .Fortran(
      "subsampling",
      input = as.integer(net),
      out = integer(n^2),
      as.integer(crit),
      as.integer(key_nodes),
      as.double(anfn),
      as.integer(numb_hidden),
      as.integer(hidden_modules),
      as.integer(n),
      as.integer(module_sizes),
      as.integer(length(module_sizes)),
      nodes_sampled = integer(n),
      edges_sampled = integer(n^2)
    )


    M <- res$edges_sampled
    M <- matrix(M, sqrt(length(M)))
    out <- igraph::graph_from_adjacency_matrix(M, weighted = TRUE)

    labs <- c("unsampled", "sampled")
    igraph::E(out)$label <- labs[igraph::E(out)$weight]
    out <- igraph::delete_edge_attr(out, "weight")

    node_labels <- labs[1 + as.integer(res$nodes_sampled > 0)]
    igraph::V(out)$label <- node_labels

    out
}



#    library(ggraph)
#    ggraph(out, layout = 'kk') +
#      geom_edge_link(aes(colour = label)) +
#      geom_node_point(aes(colour = label))


#sampled <- subgraph.edges(out, E(out)[sampled=="sampled"])
#ggraph(sampled, layout = 'kk') +
#        geom_edge_link(aes(colour = label)) +
#        geom_node_point(aes(colour = label))






R/NetGen.R
#' netgen
#'
#' Randomly generate a wide range of interaction networks
#'
#' @param net_size network size (number of nodes)
#' @param ave_module_size average module size
#' @param min_module_size cutoff for the minimum modules size
#' @param min_submod_size cutoff for submodules, used only for bipartite and tripartite networks
#' @param net_type network type, see details
#' @param ave_degree average degree of connection
#' @param rewire_prob_global probability any given edge should be rewired
#' @param rewire_prob_local probability that edges within a module should be rewire locally
#'  (within the module)
#' @param mixing_probs module probabilities for first 7 types,
#'   used for constructing mixed networks
#' @param verbose logical, default TRUE. Should a message report summary statistics?
#' @details
#' network type is one of
#' - mixed
#' - random
#' - scalefree
#' - nested
#' - bi-partite nested (or short-hand "bn")
#' - bi-partite random (or short-hand "br")
#' - tri-trophic bipartite nested-random. (Can use short-hand "ttbnr")
#' - tri-trophic bipartite nested-bipartite nested (Can use short-hand "ttbnbn")
#'
#'  **Valid Parameter Ranges**
#'
#'  Please note that not all combinations of parameters will create valid networks.
#'  If an invalid combination is requested, `netgen()` will error with an informative
#'  message.  A list of these constraints is provided below for reference.
#'
#'
#' 1. `net_size >= ave_module_size`. If `net_size = ave_module_size`` the program
#'  generates a network with a single module.
#' 2. `ave_module_size > min_module_size`
#' 3. `ave_degree >= 1`. Preferably larger than 4, to ensure single component modules.
#' 4. `rewire_prob_global = 0` produces completely uncoupled modules. To ensure a single
#'  component network use `rewire_prob_global > 0` and sufficiently large.
#' 5. `rewire_prob_local = 0` produces idealized modules.
#'  Use `rewire_prob_local > 0` to add stochasticity to the modules.
#' 6. For tripartite networks `min_module_size > min_submod_size`.
#'  This also implies `min_module_size >= 2`.
#' 7. For scalefree networks (or mixed networks involving scalefree modules)
#'  `ave_degree < min_module_size`
#' 8. For mixed networks `mixing_probs` need to sum to `1`. If the sum is larger
#'  than one, only the first types, corresponding to `sum <=1`, will be sampled.
#'
#' @importFrom igraph graph.data.frame graph_from_adjacency_matrix
#' @importFrom utils read.table
#' @return an `igraph` object
#' @export
#' @useDynLib EcoNetGen, .registration = TRUE
#' @examples
#' library(EcoNetGen)
#' \donttest{
#' set.seed(12345)
#' net <- netgen()
#' adj_plot(net)
#' }
netgen <- function(net_size = 50,
                   ave_module_size = 10,
                   min_module_size = 6,
                   min_submod_size = 1,
                   net_type = c("mixed",
                                "random",
                                "scalefree",
                                "nested",
                                "bi-partite nested",
                                "bi-partite random",
                                "tri-trophic bipartite nested-random",
                                "tri-trophic bipartite nested-bipartite nested",
                                "bn",
                                "br",
                                "tt-bn-r",
                                "tt-bn-bn"
                                ),
                   ave_degree = 5,
                   rewire_prob_global = 0.2,
                   rewire_prob_local = 0.0,
                   mixing_probs = c(0.2, 0.2, 0.2, 0.2, 0.2, 0.0 ,0.0),
                   verbose = FALSE){

  if(sum(mixing_probs) == 0){
    stop(paste("mixing_probs cannot all be zero"))
  }
  ## Normalize mixing probabilities
  mixing_probs <- mixing_probs / sum(mixing_probs)

  stopifnot(
    ave_module_size > 1,
    min_module_size > 1,
    min_submod_size >= 1,
    rewire_prob_global > 0,
    rewire_prob_global < 1,
    rewire_prob_local < 1,
    length(mixing_probs) == 7
    )

  if(ave_module_size >= net_size)
    stop(paste0("ave_module_size (",
                ave_module_size,
                ") must be less than net_size (",
                net_size, ")"))

  if(min_module_size <= min_submod_size){
    stop(paste0("min_module_size (",
               min_module_size,
               ") is not greater than min_submod_size (",
               min_submod_size, ")"))
  }

  if(ave_module_size < min_module_size){
    stop(paste0("ave_module_size (",
               ave_module_size,
               ") is not greater than or equal to min_module_size (",
               min_module_size, ")"))
  }



  net_type <- match.arg(net_type)
  net_type <- switch(net_type,
                     "mixed" = 0,
                     "random" = 1,
                     "scalefree" = 2,
                     "nested" = 3,
                     "bi-partite nested" = 41,
                     "bi-partite random" = 42,
                     "tri-trophic bipartite nested-random" = 51,
                     "tri-trophic bipartite nested-bipartite nested" = 52,
                     "bn" = 41,
                     "br" = 42,
                     "tt-bn-r" = 51,
                     "tt-bn-bn" = 52,
                     404)
  if(net_type == 404){
    stop("net_type not found")
  }

  if(net_type %in% c(0, 2)){
    if(ave_degree >= min_module_size )
    stop(paste0("Requested ave_degree (", ave_degree,
                ") must be less requested min_module_size (",
                min_module_size,
                ") for 'scalefree' or 'mixed' network types"))
  }

  n_modav <- c(net_size, ave_module_size)
  cutoffs <- c(min_module_size, min_submod_size)
  rewire_probs <- c(rewire_prob_global, rewire_prob_local)
  netgen_v1(n_modav, cutoffs, net_type, ave_degree,
            rewire_probs, mixing_probs, verbose)

}


#' netgen_v1
#'
#' netgen function
#'
#' @param n_modav network size and average module size (integer vector, length 2)
#' @param cutoffs module and submodule minimum sizes (integer vector, length 2).
#'  (submodules are used only for bipartite and tripartite networks)
#' @param net_type integer indicating type, see details
#' @param net_degree average degree of connection
#' @param net_rewire global and local  network rewiring probabilities
#' @param mod_probs module probabilities for types 1 to 51,
#'   used for constructing mixed networks, net_type = 0
#' @param verbose logical, default TRUE. Should a message report summary statistics?
#' @details
#' network type
#' - 0 = mixed
#' - 1 = random
#' - 2 = scalefree
#' - 3 = nested
#' - 41 = bi-partite nested
#' - 42 = bi-partite random
#' - 51 = tri-trophic bipartite nested-random "ttbnr"
#' - 52 = tri-trophic bipartite nested-bipartite nested "ttbnbn"
#' @importFrom igraph graph.data.frame graph_from_adjacency_matrix
#' @importFrom utils read.table
#' @return an `igraph` object
#' @useDynLib EcoNetGen, .registration = TRUE
netgen_v1 <-
  function(
           n_modav = c(50, 10),
           cutoffs = c(3, 0),
           net_type = 1,
           net_degree = 10,
           net_rewire = c(0.3,0.0),
           mod_probs = c(0.2, 0.2, 0.2, 0.2, 0.2, 0.0 ,0.0),
           verbose = FALSE) {

    stopifnot(length(n_modav) == 2,
              length(cutoffs) == 2,
              length(net_type) == 1,
              length(net_degree) == 1,
              length(mod_probs) == 7)

    res <- .Fortran(
      "subnetgen",
      output = integer(n_modav[1]^2),
      as.integer(n_modav),
      as.integer(cutoffs),
      as.integer(net_type),
      as.double(net_degree),
      as.double(net_rewire),
      as.double(mod_probs),
      modcount = integer(1L)
    )


    if(verbose){
      cluster_stats(res$output, res$modcount)
    }
    M <- res$output
    M <- matrix(M, sqrt(length(M)))
    igraph::graph_from_adjacency_matrix(M)

  }

cluster_stats <- function(M, modcount){
  n <- sqrt(length(M))
  res <- .Fortran("clusters",
                  a = as.integer(M),
                  n = as.integer(n),
                  maxsize = integer(1L),
                  icount = integer(1L))

  message(paste(
  '\nmodule count =', modcount,
  '\naverage degree =', sum(M) / n,
  '\naverage module size =', n / modcount,
  '\nnumber of components =', res$icount,
  '\nsize of largest component =', res$maxsize))
}

